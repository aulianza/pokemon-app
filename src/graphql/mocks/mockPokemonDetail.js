import { GET_POKEMON_DETAIL } from '../queries/getPokemonDetail';

export const mocksPokemonDetail = [
    {
        request: {
            query: GET_POKEMON_DETAIL,
            variables: {
                name: 'bulbasaur',
            },
        },
        result: {
            data: {
                pokemon: {
                    id: 1,
                    name: "bulbasaur",
                    sprites: {
                        front_default : 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png'
                    },
                    abilities: [
                        {
                            ability: {
                                name: "overgrow"
                            }
                        },
                        {
                            ability: {
                                name: "chlorophyll"
                            }
                        }
                    ],
                    moves: [
                        {
                            move: {
                                name: "razor-wind"
                            }
                        }
                    ],
                    types: [
                        {
                            type: {
                                name: "cut"
                            }
                        }
                    ]
                }
            },
        },
    }
]