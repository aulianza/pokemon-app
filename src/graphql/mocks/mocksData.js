import { GET_POKEMON_DETAIL } from '../queries/getPokemonDetail';
import { GET_POKEMON_LIST } from '../queries/getPokemonList';

export const mocksData = [
    {
        request: {
            query: GET_POKEMON_DETAIL,
            variables: {
                name: 'bulbasaur',
            },
        },
        result: {
            data: {
                pokemon: {
                    id: 1,
                    name: "bulbasaur",
                    sprites: {
                        front_default : 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png'
                    },
                    abilities: [
                        {
                            ability: {
                                name: "overgrow"
                            }
                        },
                        {
                            ability: {
                                name: "chlorophyll"
                            }
                        }
                    ],
                    moves: [
                        {
                            move: {
                                name: "razor-wind"
                            }
                        }
                    ],
                    types: [
                        {
                            type: {
                                name: "cut"
                            }
                        }
                    ]
                }
            },
        },
    },
    {
        request: {
            query: GET_POKEMON_LIST,
            variables: {
                limit: 2,
                offset: 0
            },
        },
        result: {
            data: {
                pokemons: {
                    results: [
                        {
                            name: "bulbasaur",
                            image: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png"
                        },
                        {
                            name: "ivysaur",
                            image: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/2.png"
                        },
                        {
                            name: "venusaur",
                            image: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/3.png"
                        }
                    ]
                }
            }
        },
    },
]