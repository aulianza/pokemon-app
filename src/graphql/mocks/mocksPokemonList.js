import { GET_POKEMON_LIST } from '../queries/getPokemonList';

export const mocksPokemonList = [
    {
        request: {
            query: GET_POKEMON_LIST,
            variables: {
                limit: 2,
                offset: 0
            },
        },
        result: {
            data: {
                pokemons: {
                    results: [
                        {
                            name: "bulbasaur",
                            image: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png"
                        },
                        {
                            name: "ivysaur",
                            image: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/2.png"
                        },
                        {
                            name: "venusaur",
                            image: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/3.png"
                        }
                    ]
                }
            }
        },
    },
]