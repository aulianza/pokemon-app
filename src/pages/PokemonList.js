import React, { useState, useRef } from "react";
import PokemonItem from "../components/PokemonItem";
import EmptyState from "../components/EmptyState";
import Button from "../components/Button";
import Loading from "../components/Loading";

import { useQuery } from "@apollo/client";
import { GET_POKEMON_LIST } from "../graphql/queries/getPokemonList";
import getOwnedPokemon from "../graphql/queries/getOwnedPokemon";
import initialDataPoint from "../graphql/operations/initialDataPoint";

const PokemonList = () => {
    const [limitState, setLimitState] = useState(initialDataPoint());

    const variables = { limit: limitState, offset: 0 };
    const { loading, error, data } = useQuery(GET_POKEMON_LIST, { variables });

    const bottomRef = useRef();
    const scrollToBottom = () => {
        if (bottomRef.current) {
            bottomRef.current.scrollIntoView({
                behavior: "smooth",
                block: "end",
                inline: "nearest",
            });
        }
    };

    const loadMoreHandler = () => {
        initialDataPoint('more');
        setLimitState(initialDataPoint());
        scrollToBottom();
    };

    const loadLessHandler = () => {
        initialDataPoint('less');
        setLimitState(initialDataPoint());
    };

    if (loading) return <Loading message="Loading..." />;
    if (error) return <EmptyState />;
    if (data) {
        const response = data.pokemons.results || [];
        const pokemonsData = response.map((pokemon) => {
            let ownedPokemon = getOwnedPokemon().filter(({ name }) => {
                return pokemon.name === name;
            });
            return Object.assign({ ...pokemon }, { owned: ownedPokemon.length });
        });

        return (
            <>
                <div className="page-header">
                    <h2>All Pokemon List</h2>
                    <code>Total: {pokemonsData.length}</code>
                </div>
                <div className="pokemon-list">
                    {pokemonsData.map((poke, index) => (
                        <PokemonItem
                            key={index}
                            poke={poke}
                            releaseAction={false}
                            currentPage="pokemonList"
                        />
                    ))}
                </div>
                <div className="pokemon-list__load__more" ref={bottomRef}>
                    {limitState > 3 ? (
                        <Button
                            title="Load Less"
                            buttonColor="btn-danger"
                            onButtonClicked={loadLessHandler}
                        />
                    ) : (
                        ""
                    )}
                    <Button
                        title="Load More"
                        buttonColor="btn-primary"
                        onButtonClicked={loadMoreHandler}
                    />
                </div>
            </>
        );
    }
};

export default PokemonList;
