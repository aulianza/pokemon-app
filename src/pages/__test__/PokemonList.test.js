import PokemonList from '../PokemonList';
import { BrowserRouter as Router } from "react-router-dom";
import { render, unmountComponentAtNode } from 'react-dom';
import { act } from "react-dom/test-utils";
import { MockedProvider } from '@apollo/client/testing';
import { mocksPokemonList } from '../../graphql/mocks/mocksPokemonList';

const data = mocksPokemonList;
let container = null;

beforeEach(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
});

afterEach(() => {
    unmountComponentAtNode(container);
    container.remove();
    container = null;
});

it("pokemon-list, snapshot", () => {
    act(() => {
        render(
            <MockedProvider mock={data}>
                <Router>
                    <PokemonList />
                </Router>
            </MockedProvider>
        ,container)
    })
    expect(container).toMatchSnapshot();
})