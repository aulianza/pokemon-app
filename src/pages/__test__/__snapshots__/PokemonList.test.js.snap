// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pokemon-list, snapshot 1`] = `
<div>
  <div
    class="css-oqpjdc"
  >
    Loading...
  </div>
</div>
`;
